// ng-zorro-antd/src/release/rate/nz-rate.component.js
import { Component, ViewEncapsulation, Input, forwardRef, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var NzRateComponent = (function () {
    function NzRateComponent() {
        this._prefixCls = 'ant-rate';
        this._innerPrefixCls = this._prefixCls + "-star";
        this._starArray = new Array();
        this._count = 5;
        this._value = 0;
        this._hoverValue = 0; // 鼠标悬浮时的星数，为正整数，和_hasHalf配合使用
        this._hasHalf = false;
        this._allowHalf = false;
        this._disabled = false;
        this._floatReg = /^\d+(\.\d+)?$/;
        // ngModel Access
        this.onChange = Function.prototype;
        this.onTouched = Function.prototype;
    }
    Object.defineProperty(NzRateComponent.prototype, "nzCount", {
        set: function (value) {
            this._count = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzRateComponent.prototype, "nzAllowHalf", {
        set: function (value) {
            if (value === '') {
                this._allowHalf = true;
            }
            else {
                this._allowHalf = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzRateComponent.prototype, "nzDefaultValue", {
        set: function (value) {
            this._value = value;
            if (this._floatReg.test(value)) {
                value += 0.5;
                this._hasHalf = true;
            }
            this._hoverValue = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzRateComponent.prototype, "nzValue", {
        get: function () {
            return this._value;
        },
        set: function (value) {
            if (this._value === value) {
                return;
            }
            this._value = value;
            if (this._floatReg.test(value)) {
                value += 0.5;
                this._hasHalf = true;
            }
            this._hoverValue = value;
        },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(NzRateComponent.prototype, "nzDisabled", {
        set: function (value) {
            this._disabled = value;
            this.setClassMap();
        },
        enumerable: true,
        configurable: true
    });
    NzRateComponent.prototype.setClassMap = function () {
        this._classMap = (_a = {},
            _a[this._prefixCls] = true,
            _a[this._prefixCls + "-disabled"] = this._disabled,
            _a);
        var _a;
    };
    NzRateComponent.prototype.setChildrenClassMap = function () {
        var index = 0;
        while (index < this._count) {
            this._starArray.push(index++);
        }
    };
    NzRateComponent.prototype._clickRate = function (e, index, isFull) {
        e.stopPropagation();
        if (this._disabled) {
            return;
        }
        this._hoverValue = this._value = index + 1;
        this._hasHalf = !isFull && this._allowHalf;
        if (this._hasHalf) {
            this._value -= 0.5;
        }
        this.onChange(this._value);
    };
    NzRateComponent.prototype._hoverRate = function (e, index, isFull) {
        e.stopPropagation();
        if (this._disabled) {
            return;
        }
        var isHalf = !isFull && this._allowHalf;
        // 如果星数一致，则不作操作，用于提高性能
        if (this._hoverValue === index + 1 && isHalf === this._hasHalf) {
            return;
        }
        this._hoverValue = index + 1;
        this._hasHalf = isHalf;
    };
    NzRateComponent.prototype._leaveRate = function (e) {
        e.stopPropagation();
        var oldVal = this._value;
        if (this._floatReg.test(oldVal)) {
            oldVal += 0.5;
            this._hasHalf = true;
        }
        this._hoverValue = oldVal;
    };
    NzRateComponent.prototype.setClasses = function (i) {
        return _a = {},
            _a[this._innerPrefixCls] = true,
            _a[this._innerPrefixCls + "-full"] = (i + 1 < this._hoverValue) || (!this._hasHalf) && (i + 1 === this._hoverValue),
            _a[this._innerPrefixCls + "-half"] = (this._hasHalf) && (i + 1 === this._hoverValue),
            _a[this._innerPrefixCls + "-active"] = (this._hasHalf) && (i + 1 === this._hoverValue),
            _a[this._innerPrefixCls + "-zero"] = (i + 1 > this._hoverValue),
            _a;
        var _a;
    };
    NzRateComponent.prototype.writeValue = function (value) {
        this.nzValue = value;
    };
    NzRateComponent.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    NzRateComponent.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    NzRateComponent.prototype.ngOnInit = function () {
        this.setClassMap();
        this.setChildrenClassMap();
    };
    return NzRateComponent;
}());
export { NzRateComponent };
NzRateComponent.decorators = [
    { type: Component, args: [{
                selector: 'nz-rate',
                encapsulation: ViewEncapsulation.None,
                template: "\n    <ul [ngClass]=\"_classMap\" (mouseleave)=\"_leaveRate($event)\">\n      <li *ngFor=\"let star of _starArray; let i = index\"\n        [ngClass]=\"setClasses(i)\"\n        (mouseover)=\"_hoverRate($event, i, true)\"\n        (click)=\"_clickRate($event, i, true)\">\n        <div class=\"ant-rate-star-first\" (mouseover)=\"_hoverRate($event, i)\" (click)=\"_clickRate($event, i)\">\n          <i class=\"anticon anticon-star\"></i></div>\n        <div class=\"ant-rate-star-second\" (mouseover)=\"_hoverRate($event, i, true)\" (click)=\"_clickRate($event, i, true)\">\n          <i class=\"anticon anticon-star\"></i></div>\n      </li>\n    </ul>\n  ",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(function () { return NzRateComponent; }),
                        multi: true
                    }
                ],
                styles: [
                    "/* stylelint-disable at-rule-empty-line-before,at-rule-name-space-after,at-rule-no-unknown */\n/* stylelint-disable declaration-bang-space-before */\n/* stylelint-disable declaration-bang-space-before */\n.ant-rate {\n  margin: 0;\n  padding: 0;\n  list-style: none;\n  font-size: 20px;\n  display: inline-block;\n  vertical-align: middle;\n}\n.ant-rate-disabled .ant-rate-star {\n  cursor: not-allowed;\n}\n.ant-rate-disabled .ant-rate-star:hover {\n  transform: scale(1);\n}\n.ant-rate-star {\n  margin: 0;\n  padding: 0;\n  display: inline-block;\n  margin-right: 8px;\n  position: relative;\n  transition: all .3s;\n  color: #e9e9e9;\n  cursor: pointer;\n}\n.ant-rate-star-first,\n.ant-rate-star-second {\n  user-select: none;\n  transition: all .3s;\n}\n.ant-rate-star:hover {\n  transform: scale(1.1);\n}\n.ant-rate-star-first {\n  position: absolute;\n  left: 0;\n  top: 0;\n  width: 50%;\n  height: 100%;\n  overflow: hidden;\n  opacity: 0;\n}\n.ant-rate-star-half .ant-rate-star-first,\n.ant-rate-star-half .ant-rate-star-second {\n  opacity: 1;\n}\n.ant-rate-star-half .ant-rate-star-first,\n.ant-rate-star-full .ant-rate-star-second {\n  color: #f5a623;\n}\n.ant-rate-star-half:hover .ant-rate-star-first,\n.ant-rate-star-full:hover .ant-rate-star-second {\n  color: #f7b84f;\n}\n.ant-rate-text {\n  margin-left: 8px;\n  vertical-align: middle;\n  display: inline-block;\n  font-size: 12px;\n}\n"
                ]
            },] },
];
/** @nocollapse */
NzRateComponent.ctorParameters = function () { return []; };
NzRateComponent.propDecorators = {
    'nzCount': [{ type: Input },],
    'nzAllowHalf': [{ type: Input },],
    'nzDefaultValue': [{ type: Input },],
    'nzDisabled': [{ type: Input },],
};
