// ng-zorro-antd/src/release/progress/nz-progress.component.js
import { Component, ViewEncapsulation, Input, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var NzProgressComponent = (function () {
    function NzProgressComponent() {
        this._statusColorMap = {
            normal: '#108ee9',
            exception: '#ff5500',
            success: '#87d068',
        };
        this._pathString = '';
        this._pathStyle = {};
        this._circleStyle = {};
        this._percent = 0;
        this._hasFormat = false;
        this._rawStatus = 'normal';
        // ngModel Access
        this.onChange = Function.prototype;
        this.onTouched = Function.prototype;
        this.nzType = 'line';
        this.nzStrokeWidth = this.nzType === 'line' ? 10 : 6;
        this.nzWidth = 132;
        this.nzShowInfo = true;
        this.nzStatus = 'normal';
        this._format = function (percent) { return percent + '%'; };
    }
    Object.defineProperty(NzProgressComponent.prototype, "_setFormat", {
        set: function (value) {
            this._format = value;
            this._hasFormat = true;
        },
        enumerable: true,
        configurable: true
    });
    NzProgressComponent.prototype.updateCircleStatus = function () {
        var circleSize = this.nzWidth || 132;
        this._circleStyle = {
            'width.px': circleSize,
            'height.px': circleSize,
            'font-size.px': circleSize * 0.16 + 6,
        };
        var radius = 50 - this.nzStrokeWidth / 2;
        var len = Math.PI * 2 * radius;
        this._pathString = 'M 50,50 m 0,-' + radius + '\n     a ' + radius + ',' + radius + ' 0 1 1 0,' + 2 * radius + '\n     a ' + radius + ',' + radius + ' 0 1 1 0,-' + 2 * radius;
        this._pathStyle = {
            'stroke-dasharray': len + 'px ' + len + 'px',
            'stroke-dashoffset': (100 - this._percent) / 100 * len + 'px',
            'transition': 'stroke-dashoffset 0.3s ease 0s, stroke 0.3s ease'
        };
    };
    NzProgressComponent.prototype.writeValue = function (value) {
        this._percent = value;
        if (this._percent === 100) {
            this.nzStatus = 'success';
        }
        else {
            this.nzStatus = this._rawStatus;
        }
        if (this.nzType === 'circle') {
            this.updateCircleStatus();
        }
    };
    NzProgressComponent.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    NzProgressComponent.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    NzProgressComponent.prototype.ngOnInit = function () {
        this._rawStatus = this.nzStatus;
    };
    return NzProgressComponent;
}());
export { NzProgressComponent };
NzProgressComponent.decorators = [
    { type: Component, args: [{
                selector: 'nz-progress',
                encapsulation: ViewEncapsulation.None,
                template: "\n    <div [ngClass]=\"'ant-progress ant-progress-status-'+nzStatus\"\n      [class.ant-progress-line]=\"nzType=='line'\"\n      [class.ant-progress-show-info]=\"nzShowInfo\"\n      [class.ant-progress-circle]=\"nzType=='circle'\">\n      <div *ngIf=\"nzType=='line'\">\n        <div class=\"ant-progress-outer\"><!----><div class=\"ant-progress-inner\"><!----><div class=\"ant-progress-bg\" [style.width.%]=\"_percent\" [style.height.px]=\"nzStrokeWidth\"></div><!----></div><!----></div><!----><span class=\"ant-progress-text\" *ngIf=\"nzShowInfo\"><ng-template [ngIf]=\"(nzStatus=='active')||(nzStatus=='normal')||(_hasFormat)\">{{_format(_percent)}}</ng-template><ng-template [ngIf]=\"(nzStatus=='exception')||(nzStatus=='success')&&(!_hasFormat)\"><i class=\"anticon\" [ngClass]=\"{'anticon-check-circle':nzStatus=='success','anticon-cross-circle':nzStatus=='exception'}\"></i></ng-template></span>\n      </div>\n      <div class=\"ant-progress-inner\" *ngIf=\"nzType=='circle'\" [ngStyle]=\"_circleStyle\">\n        <svg class=\"rc-progress-circle \" viewBox=\"0 0 100 100\" *ngIf=\"nzType=='circle'\">\n          <path class=\"rc-progress-circle-trail\" [attr.d]=\"_pathString\" stroke=\"#f3f3f3\" [attr.stroke-width]=\"nzStrokeWidth\" fill-opacity=\"0\"></path>\n          <path class=\"rc-progress-circle-path\" [attr.d]=\"_pathString\" stroke-linecap=\"round\" [attr.stroke]=\"_statusColorMap[nzStatus]\" stroke-width=\"6\" fill-opacity=\"0\" [ngStyle]=\"_pathStyle\"></path>\n        </svg>\n        <span class=\"ant-progress-text\" *ngIf=\"nzShowInfo\"><ng-template [ngIf]=\"(nzStatus=='active')||(nzStatus=='normal')||(_hasFormat)\">{{_format(_percent)}}</ng-template><ng-template [ngIf]=\"(nzStatus=='exception')||(nzStatus=='success')&&!(_hasFormat)\"><i class=\"anticon\" [ngClass]=\"{'anticon-check':nzStatus=='success','anticon-cross':nzStatus=='exception'}\"></i></ng-template></span>\n      </div>\n    </div>\n  ",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(function () { return NzProgressComponent; }),
                        multi: true
                    }
                ],
                styles: [
                    "/* stylelint-disable at-rule-empty-line-before,at-rule-name-space-after,at-rule-no-unknown */\n/* stylelint-disable declaration-bang-space-before */\n/* stylelint-disable declaration-bang-space-before */\n.ant-progress {\n  display: inline-block;\n}\n.ant-progress-line {\n  width: 100%;\n  font-size: 12px;\n  position: relative;\n}\n.ant-progress-outer {\n  display: inline-block;\n  width: 100%;\n  margin-right: 0;\n  padding-right: 0;\n}\n.ant-progress-show-info .ant-progress-outer {\n  padding-right: 2.75em;\n  margin-right: -2.75em;\n}\n.ant-progress-inner {\n  display: inline-block;\n  width: 100%;\n  background-color: #f7f7f7;\n  border-radius: 100px;\n  vertical-align: middle;\n}\n.ant-progress-circle-trail {\n  stroke: #f7f7f7;\n}\n.ant-progress-circle-path {\n  stroke: #108ee9;\n}\n.ant-progress-bg {\n  border-radius: 100px;\n  background-color: #108ee9;\n  transition: all 0.4s cubic-bezier(0.08, 0.82, 0.17, 1) 0s;\n  position: relative;\n}\n.ant-progress-text {\n  width: 2em;\n  text-align: left;\n  font-size: 1em;\n  margin-left: 0.75em;\n  vertical-align: middle;\n  display: inline-block;\n}\n.ant-progress-text .anticon {\n  font-size: 12px;\n}\n.ant-progress-status-active .ant-progress-bg:before {\n  content: \"\";\n  opacity: 0;\n  position: absolute;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  background: #fff;\n  border-radius: 10px;\n  animation: ant-progress-active 2.4s cubic-bezier(0.23, 1, 0.32, 1) infinite;\n}\n.ant-progress-status-exception .ant-progress-bg {\n  background-color: #f04134;\n}\n.ant-progress-status-exception .ant-progress-text {\n  color: #f04134;\n}\n.ant-progress-status-exception .ant-progress-circle-path {\n  stroke: #f04134;\n}\n.ant-progress-status-success .ant-progress-bg {\n  background-color: #00a854;\n}\n.ant-progress-status-success .ant-progress-text {\n  color: #00a854;\n}\n.ant-progress-status-success .ant-progress-circle-path {\n  stroke: #00a854;\n}\n.ant-progress-circle .ant-progress-inner {\n  position: relative;\n  line-height: 1;\n  background-color: transparent;\n}\n.ant-progress-circle .ant-progress-text {\n  display: block;\n  position: absolute;\n  width: 100%;\n  text-align: center;\n  line-height: 1;\n  top: 50%;\n  transform: translateY(-50%);\n  left: 0;\n  font-family: tahoma;\n  margin: 0;\n}\n.ant-progress-circle .ant-progress-text .anticon {\n  font-size: 1.16666667em;\n}\n.ant-progress-circle .ant-progress-status-exception .ant-progress-text {\n  color: #f04134;\n}\n.ant-progress-circle .ant-progress-status-success .ant-progress-text {\n  color: #00a854;\n}\n@keyframes ant-progress-active {\n  0% {\n    opacity: 0.1;\n    width: 0;\n  }\n  20% {\n    opacity: 0.5;\n    width: 0;\n  }\n  100% {\n    opacity: 0;\n    width: 100%;\n  }\n}\n"
                ]
            },] },
];
/** @nocollapse */
NzProgressComponent.ctorParameters = function () { return []; };
NzProgressComponent.propDecorators = {
    'nzType': [{ type: Input },],
    'nzStrokeWidth': [{ type: Input },],
    'nzWidth': [{ type: Input },],
    'nzShowInfo': [{ type: Input },],
    'nzStatus': [{ type: Input },],
    '_setFormat': [{ type: Input, args: ['nzFormat',] },],
};
