// ng-zorro-antd/src/release/modal/nz-modal.service.js
import { Injectable, ApplicationRef, Type, ComponentFactoryResolver } from '@angular/core';
import { NzModalComponent } from './nz-modal.component';
import { NzConfirmComponent } from './nz-confirm.component';
import { ModalOptions, ConfirmOptions } from './nz-modal-options.provider';
var NzModalService = (function () {
    function NzModalService(_appRef, _cfr) {
        this._appRef = _appRef;
        this._cfr = _cfr;
        this._modalCompFactory = this._cfr.resolveComponentFactory(NzModalComponent);
        this._confirmCompFactory = this._cfr.resolveComponentFactory(NzConfirmComponent);
    }
    NzModalService.prototype._initConfig = function (config, options) {
        if (options === void 0) { options = {}; }
        var props = {};
        var optionalParams = [
            'componentParams',
            'visible',
            'title',
            'content',
            'footer',
            'width',
            'zIndex',
            'okText',
            'cancelText',
            'style',
            'class',
            'onOk',
            'onCancel',
            'closable',
            'maskClosable',
            'wrapClassName',
            'iconType',
            'confirmType',
            'moduleWithComponentFactories'
        ];
        config = Object.assign(options, config);
        optionalParams.forEach(function (key) {
            if (config[key] !== undefined) {
                var modalKey = 'nz' + key.replace(/^\w{1}/, function (a) {
                    return a.toLocaleUpperCase();
                });
                props[modalKey] = config[key];
            }
        });
        props['onOk'] = this._getConfirmCb(props['nzOnOk']);
        props['onCancel'] = this._getConfirmCb(props['nzOnCancel']);
        // 在service模式下，不需要nzOnOk，防止触发this.nzOnOk.emit(e);
        delete props['nzOnOk'];
        delete props['nzOnCancel'];
        return props;
    };
    NzModalService.prototype._getConfirmCb = function (fn) {
        return function (_close) {
            if (fn) {
                var ret = fn();
                if (!ret) {
                    _close();
                }
                else if (ret.then) {
                    ret.then(_close);
                }
            }
            else {
                _close();
            }
        };
    };
    NzModalService.prototype._open = function (props, factory) {
        // 在body的内部最前插入一个<nz-modal></nz-modal>方便进行ApplicationRef.bootstrap
        document.body.insertBefore(document.createElement(factory.selector), document.body.firstChild);
        // document.body.appendChild(document.createElement(factory.selector));
        var customComponentFactory;
        var compRef;
        var instance;
        var subject;
        if (props['nzContent'] instanceof Type) {
            customComponentFactory = this._cfr.resolveComponentFactory(props['nzContent']);
            // 将编译出来的ngmodule中的用户component的factory作为modal内容存入
            props['nzContent'] = customComponentFactory;
        }
        compRef = this._appRef.bootstrap(factory);
        instance = compRef.instance;
        subject = instance.subject;
        ['onOk', 'onCancel'].forEach(function (eventType) {
            subject.on(eventType, function () {
                var eventHandler = props[eventType];
                if (eventHandler) {
                    eventHandler(function () {
                        instance.nzVisible = false;
                        setTimeout(function () {
                            compRef.destroy();
                        }, 200);
                    });
                }
            });
        });
        Object.assign(instance, props, {
            nzVisible: true
        });
        return subject;
    };
    // protected createComponentModule (componentType: any) {
    //   @NgModule({
    //     declarations: [ componentType ]
    //   })
    //   class CustomComponentModule {}
    //   return CustomComponentModule;
    // }
    /**
     * Open modal dialog
     */
    NzModalService.prototype.open = function (config) {
        var options = new ModalOptions();
        var props = this._initConfig(config, options);
        return this._open(props, this._modalCompFactory);
    };
    /**
     * Open confirm dialog
     */
    NzModalService.prototype._openConfirm = function (config) {
        var options = new ConfirmOptions();
        var props = this._initConfig(config, options);
        return this._open(props, this._confirmCompFactory);
    };
    /**
     * Open info dialog
     */
    NzModalService.prototype.info = function (props) {
        var config = Object.assign({}, {
            confirmType: 'info',
            iconType: 'info-circle'
        }, props);
        return this._openConfirm(config);
    };
    /**
     * Open success dialog
     */
    NzModalService.prototype.success = function (props) {
        var config = Object.assign({
            confirmType: 'success',
            iconType: 'check-circle'
        }, props);
        return this._openConfirm(config);
    };
    /**
     * Open error dialog
     */
    NzModalService.prototype.error = function (props) {
        var config = Object.assign({
            confirmType: 'error',
            iconType: 'cross-circle'
        }, props);
        return this._openConfirm(config);
    };
    /**
     * Open warning dialog
     */
    NzModalService.prototype.warning = function (props) {
        var config = Object.assign({
            confirmType: 'warning',
            iconType: 'exclamation-circle'
        }, props);
        return this._openConfirm(config);
    };
    /**
     * Open confirm dialog
     */
    NzModalService.prototype.confirm = function (props) {
        var config = Object.assign({
            confirmType: 'confirm',
            okText: '确 定',
            cancelText: '取 消'
        }, props);
        return this._openConfirm(config);
    };
    return NzModalService;
}());
export { NzModalService };
NzModalService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NzModalService.ctorParameters = function () { return [
    { type: ApplicationRef, },
    { type: ComponentFactoryResolver, },
]; };
