// ng-zorro-antd/src/release/grid/nz-col.component.js
import { Component, Input, ElementRef, HostBinding, Renderer2, Host, Optional } from '@angular/core';
import { NzRowComponent } from './nz-row.component';
var NzColComponent = (function () {
    function NzColComponent(_elementRef, _nzRow, _renderer) {
        this._elementRef = _elementRef;
        this._nzRow = _nzRow;
        this._renderer = _renderer;
        this._classList = [];
        this._prefixCls = 'ant-col';
        this._el = this._elementRef.nativeElement;
    }
    Object.defineProperty(NzColComponent.prototype, "paddingLeft", {
        get: function () {
            return this._nzRow && this._nzRow._gutter / 2;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzColComponent.prototype, "paddingRight", {
        get: function () {
            return this._nzRow && this._nzRow._gutter / 2;
        },
        enumerable: true,
        configurable: true
    });
    /** temp solution since no method add classMap to host https://github.com/angular/angular/issues/7289*/
    NzColComponent.prototype.setClassMap = function () {
        var _this = this;
        this._classList.forEach(function (_className) {
            _this._renderer.removeClass(_this._el, _className);
        });
        this._classList = [
            this.nzSpan && this._prefixCls + "-" + this.nzSpan,
            this.nzOrder && this._prefixCls + "-order-" + this.nzOrder,
            this.nzOffset && this._prefixCls + "-offset-" + this.nzOffset,
            this.nzPull && this._prefixCls + "-pull-" + this.nzPull,
            this.nzPush && this._prefixCls + "-push-" + this.nzPush,
            this.nzXs && this._prefixCls + "-xs-" + this.nzXs,
            this.nzSm && this._prefixCls + "-sm-" + this.nzSm,
            this.nzMd && this._prefixCls + "-md-" + this.nzMd,
            this.nzLg && this._prefixCls + "-lg-" + this.nzLg,
            this.nzXl && this._prefixCls + "-xl-" + this.nzXl,
        ].filter(function (item) {
            return !!item;
        });
        this._classList.forEach(function (_className) {
            _this._renderer.addClass(_this._el, _className);
        });
    };
    NzColComponent.prototype.ngOnChanges = function (changes) {
        this.setClassMap();
    };
    NzColComponent.prototype.ngOnInit = function () {
        this.setClassMap();
    };
    return NzColComponent;
}());
export { NzColComponent };
NzColComponent.decorators = [
    { type: Component, args: [{
                selector: 'nz-col',
                template: "\n    <ng-content></ng-content>\n  ",
                styles: []
            },] },
];
/** @nocollapse */
NzColComponent.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: NzRowComponent, decorators: [{ type: Optional }, { type: Host },] },
    { type: Renderer2, },
]; };
NzColComponent.propDecorators = {
    'paddingLeft': [{ type: HostBinding, args: ['style.padding-left.px',] },],
    'paddingRight': [{ type: HostBinding, args: ['style.padding-right.px',] },],
    'nzSpan': [{ type: Input },],
    'nzOrder': [{ type: Input },],
    'nzOffset': [{ type: Input },],
    'nzPush': [{ type: Input },],
    'nzPull': [{ type: Input },],
    'nzXs': [{ type: Input },],
    'nzSm': [{ type: Input },],
    'nzMd': [{ type: Input },],
    'nzLg': [{ type: Input },],
    'nzXl': [{ type: Input },],
};
