// ng-zorro-antd/src/release/select/nz-option.component.js
import { Component, ViewEncapsulation, Input } from '@angular/core';
import { NzSelectComponent } from './nz-select.component';
var NzOptionComponent = (function () {
    function NzOptionComponent(_nzSelect) {
        this._nzSelect = _nzSelect;
        this._disabled = false;
        this._nzSelect.addOption(this);
    }
    Object.defineProperty(NzOptionComponent.prototype, "nzValue", {
        get: function () {
            return this._value;
        },
        set: function (value) {
            if (this._value === value) {
                return;
            }
            this._value = value;
        },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(NzOptionComponent.prototype, "nzLabel", {
        get: function () {
            return this._label;
        },
        set: function (value) {
            if (this._label === value) {
                return;
            }
            this._label = value;
        },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(NzOptionComponent.prototype, "nzDisabled", {
        get: function () {
            return this._disabled;
        },
        set: function (value) {
            this._disabled = value;
        },
        enumerable: true,
        configurable: true
    });
    ;
    NzOptionComponent.prototype.ngOnDestroy = function () {
        this._nzSelect.removeOption(this);
    };
    return NzOptionComponent;
}());
export { NzOptionComponent };
NzOptionComponent.decorators = [
    { type: Component, args: [{
                selector: 'nz-option',
                encapsulation: ViewEncapsulation.None,
                template: "\n    <ng-content></ng-content>\n  ",
                styles: []
            },] },
];
/** @nocollapse */
NzOptionComponent.ctorParameters = function () { return [
    { type: NzSelectComponent, },
]; };
NzOptionComponent.propDecorators = {
    'nzValue': [{ type: Input },],
    'nzLabel': [{ type: Input },],
    'nzDisabled': [{ type: Input },],
};
