// ng-zorro-antd/src/release/timeline/nz-timeline-item.component.js
import { Component, Input, ContentChild, ViewEncapsulation } from '@angular/core';
var NzTimelineItemComponent = (function () {
    function NzTimelineItemComponent() {
        this.itemHeadClass = { 'ant-timeline-item-head-blue': true };
        this._color = 'blue';
        this._custom = false;
        this._lastItem = false;
    }
    Object.defineProperty(NzTimelineItemComponent.prototype, "nzColor", {
        get: function () {
            return this._color;
        },
        set: function (color) {
            this._color = color;
            if (color === 'green') {
                this.itemHeadClass['ant-timeline-item-head-green'] = true;
            }
            else if (color === 'red') {
                this.itemHeadClass['ant-timeline-item-head-red'] = true;
            }
            else {
                this.itemHeadClass['ant-timeline-item-head-blue'] = true;
            }
        },
        enumerable: true,
        configurable: true
    });
    NzTimelineItemComponent.prototype.ngOnInit = function () {
        if (this._customContent) {
            this._custom = true;
        }
    };
    return NzTimelineItemComponent;
}());
export { NzTimelineItemComponent };
NzTimelineItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'nz-timeline-item',
                encapsulation: ViewEncapsulation.None,
                template: "\n    <li\n      class=\"ant-timeline-item\"\n      [class.ant-timeline-item-last]=\"_lastItem\">\n      <div class=\"ant-timeline-item-tail\"></div>\n      <div class=\"ant-timeline-item-head\"\n        [class.ant-timeline-item-head-custom]=\"_custom\"\n        [ngClass]=\"itemHeadClass\">\n        <ng-template [ngTemplateOutlet]=\"_customContent\">\n        </ng-template>\n      </div>\n      <div class=\"ant-timeline-item-content\">\n        <ng-content></ng-content>\n      </div>\n    </li>",
                styles: []
            },] },
];
/** @nocollapse */
NzTimelineItemComponent.ctorParameters = function () { return []; };
NzTimelineItemComponent.propDecorators = {
    '_customContent': [{ type: ContentChild, args: ['custom',] },],
    'nzColor': [{ type: Input },],
};
