// ng-zorro-antd/src/release/slider/nz-slider-step.component.js
import { Component, Input, ViewEncapsulation } from '@angular/core';
var NzSliderStepComponent = (function () {
    function NzSliderStepComponent() {
        // Dynamic properties
        this.nzLowerBound = null;
        this.nzUpperBound = null;
    }
    NzSliderStepComponent.prototype.ngOnChanges = function (changes) {
        if (changes.nzLowerBound || changes.nzUpperBound) {
            this.togglePointActive();
        }
    };
    NzSliderStepComponent.prototype.ngOnInit = function () {
        var orient = this.nzVertical ? 'bottom' : 'left', prefixCls = this.nzPrefixCls;
        this.attrs = this.nzMarksArray.map(function (mark) {
            var value = mark.value, offset = mark.offset;
            return {
                id: value,
                value: value,
                offset: offset,
                style: (_a = {},
                    _a[orient] = offset + "%",
                    _a),
                classes: (_b = {},
                    _b[prefixCls + "-dot"] = true,
                    _b[prefixCls + "-dot-active"] = false,
                    _b)
            };
            var _a, _b;
        });
        this.togglePointActive();
    };
    NzSliderStepComponent.prototype.trackById = function (index, attr) {
        return attr.id;
    };
    NzSliderStepComponent.prototype.togglePointActive = function () {
        var _a = this, nzPrefixCls = _a.nzPrefixCls, attrs = _a.attrs, nzLowerBound = _a.nzLowerBound, nzUpperBound = _a.nzUpperBound, nzIncluded = _a.nzIncluded;
        if (attrs && nzLowerBound !== null && nzUpperBound !== null) {
            attrs.forEach(function (attr) {
                var value = attr.value, isActive = (!nzIncluded && value === nzUpperBound) ||
                    (nzIncluded && value <= nzUpperBound && value >= nzLowerBound);
                attr.classes[nzPrefixCls + "-dot-active"] = isActive;
            });
        }
    };
    return NzSliderStepComponent;
}());
export { NzSliderStepComponent };
NzSliderStepComponent.decorators = [
    { type: Component, args: [{
                selector: 'nz-slider-step',
                encapsulation: ViewEncapsulation.None,
                template: "\n    <div class=\"{{nzPrefixCls}}-step\">\n      <span *ngFor=\"let attr of attrs; trackBy: trackById\" [ngClass]=\"attr.classes\" [ngStyle]=\"attr.style\"></span>\n    </div>\n  "
            },] },
];
/** @nocollapse */
NzSliderStepComponent.ctorParameters = function () { return []; };
NzSliderStepComponent.propDecorators = {
    'nzLowerBound': [{ type: Input },],
    'nzUpperBound': [{ type: Input },],
    'nzPrefixCls': [{ type: Input },],
    'nzVertical': [{ type: Input },],
    'nzMarksArray': [{ type: Input },],
    'nzIncluded': [{ type: Input },],
};
