// ng-zorro-antd/src/release/slider/nz-slider-track.component.js
import { Component, Input, ViewEncapsulation } from '@angular/core';
var NzSliderTrackComponent = (function () {
    function NzSliderTrackComponent() {
        this.style = {};
    }
    NzSliderTrackComponent.prototype.ngOnChanges = function (changes) {
        var _a = this, nzOffset = _a.nzOffset, nzLength = _a.nzLength, nzIncluded = _a.nzIncluded, nzVertical = _a.nzVertical, style = _a.style;
        if (changes.nzIncluded) {
            style.visibility = nzIncluded ? 'visible' : 'hidden';
        }
        if (changes.nzVertical || changes.nzOffset || changes.nzLength) {
            if (nzVertical) {
                style.bottom = nzOffset + "%";
                style.height = nzLength + "%";
            }
            else {
                style.left = nzOffset + "%";
                style.width = nzLength + "%";
            }
        }
    };
    return NzSliderTrackComponent;
}());
export { NzSliderTrackComponent };
NzSliderTrackComponent.decorators = [
    { type: Component, args: [{
                selector: 'nz-slider-track',
                encapsulation: ViewEncapsulation.None,
                template: "\n    <div [class]=\"nzClassName\" [ngStyle]=\"style\"></div>\n  "
            },] },
];
/** @nocollapse */
NzSliderTrackComponent.ctorParameters = function () { return []; };
NzSliderTrackComponent.propDecorators = {
    'nzOffset': [{ type: Input },],
    'nzLength': [{ type: Input },],
    'nzClassName': [{ type: Input },],
    'nzVertical': [{ type: Input },],
    'nzIncluded': [{ type: Input },],
};
