// ng-zorro-antd/src/release/form/nz-form-explain.directive.js
import { Component, HostBinding, ViewEncapsulation } from '@angular/core';
import { NzFormItemDirective } from './nz-form-item.directive';
var NzFormExplainComponent = (function () {
    function NzFormExplainComponent(_nzFormItem) {
        this._nzFormItem = _nzFormItem;
    }
    NzFormExplainComponent.prototype.ngOnDestroy = function () {
        this._nzFormItem.disableHelp();
    };
    NzFormExplainComponent.prototype.ngOnInit = function () {
        this._nzFormItem.enableHelp();
    };
    return NzFormExplainComponent;
}());
export { NzFormExplainComponent };
NzFormExplainComponent.decorators = [
    { type: Component, args: [{
                selector: '[nz-form-explain]',
                encapsulation: ViewEncapsulation.None,
                template: "\n    <ng-content></ng-content>\n  ",
                styles: []
            },] },
];
/** @nocollapse */
NzFormExplainComponent.ctorParameters = function () { return [
    { type: NzFormItemDirective, },
]; };
NzFormExplainComponent.propDecorators = {
    'true': [{ type: HostBinding, args: ["class.ant-form-explain",] },],
};
