// ng-zorro-antd/src/release/carousel/nz-carousel-content.directive.js
import { Directive, HostBinding } from '@angular/core';
var NzCarouselContentDirective = (function () {
    function NzCarouselContentDirective() {
        this.width = 0;
        this.isActive = false;
        this.left = null;
        this.top = null;
        this.fadeMode = false;
    }
    Object.defineProperty(NzCarouselContentDirective.prototype, "setActiveClass", {
        get: function () {
            return this.isActive === true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzCarouselContentDirective.prototype, "setWidth", {
        get: function () {
            return this.width;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzCarouselContentDirective.prototype, "setLeft", {
        get: function () {
            return this.left;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzCarouselContentDirective.prototype, "setTop", {
        get: function () {
            return this.top;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzCarouselContentDirective.prototype, "setPosition", {
        get: function () {
            if (this.fadeMode) {
                return 'relative';
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzCarouselContentDirective.prototype, "setOpacity", {
        get: function () {
            if (this.fadeMode) {
                return this.isActive ? 1 : 0;
            }
        },
        enumerable: true,
        configurable: true
    });
    return NzCarouselContentDirective;
}());
export { NzCarouselContentDirective };
NzCarouselContentDirective.decorators = [
    { type: Directive, args: [{
                selector: '[nz-carousel-content]',
            },] },
];
/** @nocollapse */
NzCarouselContentDirective.ctorParameters = function () { return []; };
NzCarouselContentDirective.propDecorators = {
    'true': [{ type: HostBinding, args: ['class.slick-slide',] },],
    'setActiveClass': [{ type: HostBinding, args: ['class.slick-active',] },],
    'setWidth': [{ type: HostBinding, args: ['style.width.px',] },],
    'setLeft': [{ type: HostBinding, args: ['style.left.px',] },],
    'setTop': [{ type: HostBinding, args: ['style.top.px',] },],
    'setPosition': [{ type: HostBinding, args: ['style.position',] },],
    'setOpacity': [{ type: HostBinding, args: ['style.opacity',] },],
};
