// ng-zorro-antd/src/release/button/nz-button-group.component.js
import { Component, Input, ViewEncapsulation, ViewChild } from '@angular/core';
var NzButtonGroupComponent = (function () {
    function NzButtonGroupComponent() {
        this._prefixCls = 'ant-btn-group';
        this._sizeMap = { large: 'lg', small: 'sm' };
    }
    Object.defineProperty(NzButtonGroupComponent.prototype, "nzSize", {
        get: function () {
            return this._size;
        },
        set: function (value) {
            this._size = value;
        },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(NzButtonGroupComponent.prototype, "_classMap", {
        get: function () {
            return _a = {},
                _a[this._prefixCls] = true,
                _a[this._prefixCls + "-" + this._sizeMap[this.nzSize]] = this._sizeMap[this.nzSize],
                _a;
            var _a;
        },
        enumerable: true,
        configurable: true
    });
    ;
    NzButtonGroupComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        /** trim text node between button */
        Array.from(this._groupWrapper.nativeElement.childNodes).forEach(function (node) {
            if (node.nodeType === 3) {
                _this._groupWrapper.nativeElement.removeChild(node);
            }
        });
    };
    return NzButtonGroupComponent;
}());
export { NzButtonGroupComponent };
NzButtonGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'nz-button-group',
                encapsulation: ViewEncapsulation.None,
                template: "\n    <div [ngClass]=\"_classMap\" #groupWrapper>\n      <ng-content></ng-content>\n    </div>\n  ",
                styles: []
            },] },
];
/** @nocollapse */
NzButtonGroupComponent.ctorParameters = function () { return []; };
NzButtonGroupComponent.propDecorators = {
    '_groupWrapper': [{ type: ViewChild, args: ['groupWrapper',] },],
    'nzSize': [{ type: Input },],
};
