// ng-zorro-antd/src/release/tabs/nz-tab-label.directive.js
import { Directive, ElementRef, HostBinding, Input } from '@angular/core';
var NzTabLabelDirective = (function () {
    function NzTabLabelDirective(elementRef) {
        this.elementRef = elementRef;
        this.disabled = false;
    }
    NzTabLabelDirective.prototype.getOffsetLeft = function () {
        return this.elementRef.nativeElement.offsetLeft;
    };
    NzTabLabelDirective.prototype.getOffsetWidth = function () {
        return this.elementRef.nativeElement.offsetWidth;
    };
    NzTabLabelDirective.prototype.getOffsetTop = function () {
        return this.elementRef.nativeElement.offsetTop;
    };
    NzTabLabelDirective.prototype.getOffsetHeight = function () {
        return this.elementRef.nativeElement.offsetHeight;
    };
    return NzTabLabelDirective;
}());
export { NzTabLabelDirective };
NzTabLabelDirective.decorators = [
    { type: Directive, args: [{
                selector: '[nz-tab-label]'
            },] },
];
/** @nocollapse */
NzTabLabelDirective.ctorParameters = function () { return [
    { type: ElementRef, },
]; };
NzTabLabelDirective.propDecorators = {
    'true': [{ type: HostBinding, args: ['class.ant-tabs-tab',] },],
    'disabled': [{ type: Input }, { type: HostBinding, args: ['class.ant-tabs-tab-disabled',] },],
};
